core/src/Kokkos_Parallel.hpp

```diff
@@ -137,9 +137,9 @@ inline void parallel_for(const std::string& str, const ExecPolicy& policy,
   ExecPolicy inner_policy = policy;
   Kokkos::Tools::Impl::begin_parallel_for(inner_policy, functor, str, kpID);
 
-  Kokkos::Impl::shared_allocation_tracking_disable();
-  Impl::ParallelFor<FunctorType, ExecPolicy> closure(functor, inner_policy);
-  Kokkos::Impl::shared_allocation_tracking_enable();
+  auto closure = Kokkos::Impl::with_shared_allocation_tracking_disabled([&]() {
+    return Impl::ParallelFor<FunctorType, ExecPolicy>(functor, inner_policy);
+  });
 
   closure.execute();
```

Comment:
I am torn which is better.
At the moment I am thinking this is slightly easier to reason about.
```suggestion
  {
  [[maybe_unused]] auto guard = SharedAllocationDisableTrackingGuard{};
  auto closure =  Impl::ParallelFor<FunctorType, ExecPolicy>(functor, inner_policy);
  closure.execute();
  }
```

Reply:
I adopted @masterleinad 's suggestion which will hopefully be a bit easier to understand